ant/ant.go

```diff
@@ -72,6 +88,71 @@ func clearPorts(config AntConfig) error {
 	return nil
 }
 
+// upgraderThread is a goroutine that waits `UpgradeDelay` and upgrades to the
+// next version in the ant's `UpgradePath`. Once the final version is reached,
+// upgrading is stopped.
+func (a *Ant) upgraderThread() {
+	err := a.tg.Add()
+	if err != nil {
+		return
+	}
+	defer a.tg.Done()
+
+	for _, version := range a.Config.UpgradePath {
+		select {
+		case <-time.After(time.Second * time.Duration(a.Config.UpgradeDelay)):
+		case <-a.tg.StopChan():
+			return
+		}
+
+		log.Printf("upgrading %v to %v...\n", a.Config.Name, version)
+
+		stopSiad(a.Config.APIAddr, a.siad.Process)
+		a.jr.Stop()
+
+		newSiadPath := path.Join(a.Config.UpgradeDir, fmt.Sprintf("%v-%v-%v", version, runtime.GOOS, runtime.GOARCH), "siad")
+
+		siad, err := newSiad(newSiadPath, a.Config.SiaDirectory, a.Config.APIAddr, a.Config.RPCAddr, a.Config.HostAddr)
+		if err != nil {
+			log.Printf("error starting siad after upgrade: %v\n", err)
+			continue
+		}
+
+		j := &jobRunner{
+			client:         api.NewClient(a.Config.APIAddr, ""),
+			siaDirectory:   a.Config.SiaDirectory,
+			walletPassword: a.jr.walletPassword,
+		}
+
+		err = j.client.Post("/wallet/unlock", fmt.Sprintf("encryptionpassword=%s&dictionary=%s", a.jr.walletPassword, "english"), nil)
+		if err != nil {
+			continue
+		}
+
+		for _, job := range a.Config.Jobs {
+			switch job {
+			case "miner":
+				go j.blockMining()
+			case "host":
+				go j.jobHost()
```

Comment:
this seems to cause problems. I don't think the host job needs to be explicitly restarted. (Well, it probably should be, but it needs to detect that it was restarted.)